<!-- PROJECT LOGO -->
<br />
<p align="center">
  <h3 align="center">Neurosymbolic Neuronal Network Modeling</h3>

  <p align="center">
    Modeling and Predicting Neuronal Network Behavior Using Gene Expression Data
  </p>
</p>

<!-- TABLE OF CONTENTS -->
<details open="open">
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
      <ul>
        <li><a href="#built-with">Built With</a></li>
      </ul>
    </li>
    <li>
      <a href="#getting-started">Getting Started</a>
      <ul>
        <li><a href="#dependencies">Dependencies</a></li>
        <li><a href="#installation">Installation</a></li>
      </ul>
    </li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="#roadmap">Roadmap</a></li>
    <li><a href="#contributing">Contributing</a></li>
    <li><a href="#license">License</a></li>
    <li><a href="#authors">Authors</a></li>
    <li><a href="#acknowledgements">Acknowledgements</a></li>
  </ol>
</details>

<!-- ABOUT THE PROJECT -->
## About The Project

This project aims to model and predict neuronal network behavior by analyzing the gene expression of individual neurons, characterized by their anatomical and functional properties. The dataset used for this project can be found [here](https://www.refine.bio/experiments/SRP094496/correlating-anatomy-and-function-with-gene-expression-in-individual-neurons-by-combining-in-vivo-labeling-patch-clamp-and-single-cell-rna-seq).

Research Question: "How can the transcriptional profiles of individual neurons, characterized by their anatomical and functional properties, be used to model and predict neuronal network behavior in a neurosymbolic framework?"

<!-- GETTING STARTED -->
## Getting Started

### Dependencies

* Python 3.11
* Pandas
* NumPy
* scikit-learn
* TensorFlow or PyTorch

### Installation

1. Clone the repository to your local machine.
2. Install the required dependencies.
3. Download the [dataset](https://www.refine.bio/experiments/SRP094496/correlating-anatomy-and-function-with-gene-expression-in-individual-neurons-by-combining-in-vivo-labeling-patch-clamp-and-single-cell-rna-seq) and place it in the appropriate directory.

<!-- USAGE EXAMPLES -->
## Usage

Detailed instructions on how to run the code and use the dataset will be provided.

<!-- ROADMAP -->
## Roadmap

1. Data Preprocessing and Exploration
2. Feature Engineering
3. Model Training and Evaluation
4. Interpretation and Analysis

<!-- LICENSE -->
## License

Distributed under the MIT License. See `LICENSE` for more information.

<!-- Authors -->
## Authors

Ethan Fan - [Github](https://github.com/Ethan12103)

Matheus Kunzler Maldaner - [Github](https://github.com/matheusmaldaner)

Nathan Gilman - [Github](https://github.com/npgilman)

Rama Janco - [Github](https://github.com/ramawama)

<!-- ACKNOWLEDGEMENTS -->
## Acknowledgements

* Kiley Graim for instructing the course.
* Pfeffer CK and Beltramo R for authoring the original research and dataset.
* The Sequence Read Archive (SRA) for hosting the source repository.
* refine.bio and the Childhood Cancer Data Lab (CCDL) for providing a uniformly processed and normalized version of the dataset.
* Alex's Lemonade Stand Foundation for powering the refine.bio project.
* Casey S. Greene, Dongbo Hu, Richard W. W. Jones, Stephanie Liu, David S. Mejia, Rob Patro, Stephen R. Piccolo, Ariel Rodriguez Romero, Hirak Sarkar, Candace L. Savonen, Jaclyn N. Taroni, William E. Vauclain, Deepashree Venkatesh Prasad, Kurt G. Wheeler for their contributions to refine.bio.


## Thank you


